File: February/CPP/Cheapest Flights Within K Stops/Cheapest Flights Within K Stops.cpp
class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) {
        unordered_map<int,vector<pair<int,int>>> adj;
        for(auto& x: flights) adj[x[0]].push_back({x[1],x[2]});
        vector<int> dist(n,INT_MAX);
        dist[src] = 0;
        queue<pair<int,int>> q;
        q.push({src,0});
        int stops = 0;
        while(!q.empty() && stops<=k){
            int sz = q.size();
            while(sz-- >0){
                auto [node, distance] = q.front();
                q.pop();
                if(!adj.count(node)) continue;
                for(auto& [neighbour,price]: adj[node]){
                    if(price+distance>=dist[neighbour]) continue;
                    dist[neighbour] = price+distance;
                    q.push({neighbour, dist[neighbour]});
                }
            }
            stops++;
        }
        return dist[dst] == INT_MAX?-1: dist[dst];
    }
};

File: February/Python/Cheapest Flights Within K Stops/Cheapest Flights Within K Stops.py
class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, k: int) -> int:
        adj = defaultdict(list)
        for l_src, l_dest, l_price in flights:
            adj[l_src].append((l_dest, l_price))

        dist = [float('inf')] * n
        dist[src] = 0

        q = [(src, 0)]
        stops = 0

        while stops <= k and q:
            size = len(q)
            for _ in range(size):
                c_src, src_price = q.pop(0)
                for c_dest, dest_price in adj[c_src]:
                    if src_price + dest_price >= dist[c_dest]:
                        continue
                    dist[c_dest] = src_price + dest_price
                    q.append((c_dest, dist[c_dest]))
            stops += 1

        return dist[dst] if  dist[dst] != float('inf') else -1
